### PR TITLE
stop AST manipulation

### PR DESCRIPTION
**Pull request by pq**
Some work to move away from deprecated APIs.

@scheglov: could you advise on the best approach for `unnecessary_const`?

/cc @scheglov

### CODE REVIEW

File: lib/src/rules/unnecessary_const.dart [hunk 1]
```diff
@@ -82,6 +82,8 @@ class _Visitor extends SimpleAstVisitor {
   void _visitTypedLiteral(TypedLiteral node) {
     if (node.constKeyword == null) return;
 
+    // todo (pq): update to use linter context API
```
Comment (bwilkerson) [inline lib/src/rules/unnecessary_const.dart:85]: I think you should be able to ask `node.inConstantContext`.

Comment (scheglov) [reply lib/src/rules/unnecessary_const.dart:85]: Yes, it looks that it will work.
One complication is that it declared in `ExpressionImpl`.

Comment (bwilkerson) [reply lib/src/rules/unnecessary_const.dart:85]: Should we make it public? Do we need to wrap it in `LintContext`?

Comment (scheglov) [reply lib/src/rules/unnecessary_const.dart:85]: We don't need manipulations with AST to implement it, so I think we can make it public.

Do you like its name? Should it be `isInConstantContext`? OTOH we also have `inGetterContext`, so it seems consistent. Just don't know if consistently good, or consistently bad :-)

Comment (pq) [reply lib/src/rules/unnecessary_const.dart]: Thanks!  Seems like this could go in `LinterContext`?

**Comment by pq** [comment]
@scheglov, @bwilkerson: not pretty, but better?

File: lib/src/rules/unnecessary_const.dart (hunk 1 above)
Comment (pq) [reply lib/src/rules/unnecessary_const.dart]: (whoops.  missed your earlier comment.)

> We don't need manipulations with AST to implement it, so I think we can make it public.

I can't speak to consistency.  Either work for me.  I guess I'd lean towards `inConstantContext`.  If nothing else, it's shorter!

File: lib/src/rules/unnecessary_const.dart [hunk 2]
```diff
@@ -60,13 +62,9 @@ class _Visitor extends SimpleAstVisitor {
...
-    // remove keyword and check if there's const error
-    final oldKeyword = node.keyword;
-    node.keyword = null;
-    final isConstWithoutKeyword = node.isConst;
-    node.keyword = oldKeyword;
-
-    if (isConstWithoutKeyword && node.keyword?.type == Keyword.CONST) {
+    // todo (pq): remove cast when `inConstantContext` is public on Element
+    var inConstContext = (node as ExpressionImpl).inConstantContext;
+    if (inConstContext && node.keyword?.type == Keyword.CONST) {
```
Comment (scheglov) [inline lib/src/rules/unnecessary_const.dart:67]: I would have switched the order, and checked `node.keyword?.type == Keyword.CONST` first (as you have in the first statement of this method). It is cheaper.

File: lib/src/rules/unnecessary_const.dart [hunk 3]
```diff
@@ -82,13 +80,9 @@ class _Visitor extends SimpleAstVisitor {
...
-    // remove keyword and check if there's const error
-    final oldKeyword = node.constKeyword;
-    node.constKeyword = null;
-    final isConstWithoutKeyword = node.isConst;
-    node.constKeyword = oldKeyword;
-
-    if (isConstWithoutKeyword && node.constKeyword?.type == Keyword.CONST) {
+    // todo (pq): remove cast when `inConstantContext` is public on Element
+    var inConstContext = (node as ExpressionImpl).inConstantContext;
+    if (inConstContext && node.constKeyword?.type == Keyword.CONST) {
```
Comment (scheglov) [inline lib/src/rules/unnecessary_const.dart:85]: Can `constKeyword` be anything else?

File: lib/src/rules/unnecessary_const.dart [hunk 4]
```diff
@@ -60,13 +62,9 @@ class _Visitor extends SimpleAstVisitor {
...
     if (node.keyword == null) return;
 
-    // remove keyword and check if there's const error
-    final oldKeyword = node.keyword;
-    node.keyword = null;
-    final isConstWithoutKeyword = node.isConst;
-    node.keyword = oldKeyword;
-
-    if (isConstWithoutKeyword && node.keyword?.type == Keyword.CONST) {
+    // todo (pq): remove cast when `inConstantContext` is public on Element
```
Comment (bwilkerson) [inline lib/src/rules/unnecessary_const.dart:65]: "Element" --> "Expression"

**Review by scheglov — APPROVED** [review]

File: lib/src/rules/unnecessary_const.dart (hunk 2 above)
Comment (bwilkerson) [inline lib/src/rules/unnecessary_const.dart:67]: Consider moving `node.keyword?.type == Keyword.CONST` to line 63. It's a less expensive check so it should be done first, and there's really no reason to check that the keyword is non-null in t addition to checking that it is the right keyword.

**Review by bwilkerson — APPROVED** [review]
> Should it be `isInConstantContext`?

No, I don't think the "is" adds any value in this case, and it's more consistent.

(Not sure why, but I can't seem to reply to that question in-line.)

File: lib/src/rules/unnecessary_const.dart (hunk 3 above)
Comment (bwilkerson) [reply lib/src/rules/unnecessary_const.dart:85]: There's already a check on line 81.

File: lib/src/rules/unnecessary_const.dart (hunk 2 above)
Comment (pq) [reply lib/src/rules/unnecessary_const.dart]: Good catch.  Konstantin had a similar suggestion. Updated.

**Pull request merged**